SpatialGDK/Build/Programs/Improbable.Unreal.Scripts/DeploymentLauncher/DeploymentLauncher.cs

```diff
@@ -0,0 +1,476 @@
+using System.Linq;
+using System.Net;
+using System.Security.Cryptography;
+using Google.LongRunning;
+using Google.Protobuf.WellKnownTypes;
+using Improbable.SpatialOS.Deployment.V1Alpha1;
+using Improbable.SpatialOS.PlayerAuth.V2Alpha1;
+using Improbable.SpatialOS.Snapshot.V1Alpha1;
+using Newtonsoft.Json.Linq;
+
+namespace Improbable
+{
+    internal class DeploymentLauncher
+    {
+        private const string SIM_PLAYER_DEPLOYMENT_TAG = "simulated_players";
+        private const string DEPLOYMENT_LAUNCHED_BY_LAUNCHER_TAG = "unreal_deployment_launcher";
+
+        private const string CoordinatorWorkerName = "SimulatedPlayerCoordinator";
+
+        private static string UploadSnapshot(SnapshotServiceClient client, string snapshotPath, string projectName,
+            string deploymentName)
+        {
+            Console.WriteLine($"Uploading {snapshotPath} to project {projectName}");
+
+            // Read snapshot.
+            var bytes = File.ReadAllBytes(snapshotPath);
+
+            if (bytes.Length == 0)
+            {
+                Console.Error.WriteLine($"Unable to load {snapshotPath}. Does the file exist?");
+                return string.Empty;
+            }
+
+            // Create HTTP endpoint to upload to.
+            var snapshotToUpload = new Snapshot
+            {
+                ProjectName = projectName,
+                DeploymentName = deploymentName
+            };
+
+            using (var md5 = MD5.Create())
+            {
+                snapshotToUpload.Checksum = Convert.ToBase64String(md5.ComputeHash(bytes));
+                snapshotToUpload.Size = bytes.Length;
+            }
+
+            var uploadSnapshotResponse =
+                client.UploadSnapshot(new UploadSnapshotRequest { Snapshot = snapshotToUpload });
+            snapshotToUpload = uploadSnapshotResponse.Snapshot;
+
+            // Upload content.
+            var httpRequest = WebRequest.Create(uploadSnapshotResponse.UploadUrl) as HttpWebRequest;
+            httpRequest.Method = "PUT";
+            httpRequest.ContentLength = snapshotToUpload.Size;
+            httpRequest.Headers.Set("Content-MD5", snapshotToUpload.Checksum);
+
+            using (var dataStream = httpRequest.GetRequestStream())
+            {
+                dataStream.Write(bytes, 0, bytes.Length);
+            }
+
+            // Block until we have a response.
+            httpRequest.GetResponse();
+
+            // Confirm that the snapshot was uploaded successfully.
+            var confirmUploadResponse = client.ConfirmUpload(new ConfirmUploadRequest
+            {
+                DeploymentName = snapshotToUpload.DeploymentName,
+                Id = snapshotToUpload.Id,
+                ProjectName = snapshotToUpload.ProjectName
+            });
+
+            return confirmUploadResponse.Snapshot.Id;
+        }
+
+        private static int CreateDeployment(string[] args)
+        {
+            bool launchSimPlayerDeployment = args.Length == 11;
```

Comment:
Looks like we only support 7 or 11 params, should we do some error checking if an incorrect number is passed in?

Reply:
Oh right, this isn't main. Don't mind me.